### ISSUE TITLE
Azure Chatbot, Webchat Widget accessibility keyboard navigation and screen reader focus order is not logical.

### ISSUE THREAD

**Issue by jason-walgreens** [microsoft/BotFramework-WebChat#4201]
# 🚨 The issue tracker is not for implementation questions 🚨

# Please view our [Technical Support Guide](https://github.com/microsoft/BotFramework-WebChat/tree/main/docs/TECHNICAL_SUPPORT_GUIDE.md) before filing a new issue.

This repo focuses on the development of Web Chat, a client/channel for Bot Framework. If you have questions on implementation or about other features of Bot Framework, please see the below list to direct your questions to the correct repo or site.

| Topic                       | Description                                                  | Link                                                    |
| --------------------------- | ------------------------------------------------------------ | ------------------------------------------------------- |
| Web Chat 'How to' Questions | Ask implementation questions about Web Chat                  | https://stackoverflow.com/questions/tagged/web-chat     |
| Bot Framework Questions     | Ask implementation questions related to the BotFramework SDK | https://stackoverflow.com/questions/tagged/botframework |
| Bot Builder                 | A comprehensive list of Bot Framework SDKs and tools         | https://github.com/microsoft/BotBuilder                 |

[Question]

Azure Chatbot, Webchat Widget accessibility keyboard navigation and screen reader focus order is not logical.

Need clarification for screen reader focus order and key navigation when new buttons are available.

Ex:
i) When new buttons are available, it requires users to press "enter" and then "tab" key to access the buttons. The screen reader does not read this direction all the time. 
ii) When a set of buttons are available on the screen, focus still trapped on the first set of buttons.

**Comment by sknaappilaswd** [comment]
I have also observed the same malfunctions when there are multiple sets of buttons available. The focus stays only in the first set of buttons (when pressing UP or DOWN arrow keys).

![image](https://user-images.githubusercontent.com/79024237/157628588-34fe401e-eb53-417e-93a0-a72152319b29.png)

**Comment by compulim** [comment]
@jason-walgreens

For your first point, when a message with buttons arrive (e.g. Adaptive Cards), the screen reader should narrate: "Bot said: <message body>. Message is interactive. Press shift tab key 2 to 3 times to switch to the chat history. Then click on the message to interact." Are you seeing some messages with buttons arrived without this narratives? If yes, could you record a short video showing this behavior?

For your second point, according to [keyboard navigation pattern in WAI-ARIA](https://www.w3.org/TR/wai-aria-practices-1.2/#kbd_general_within), we put the focus on the element which the user is focused last time. Thus, we are not moving the focus to the new message. Excerpt from the page:

> The element that had focus the last time the composite contained focus. Or, if the composite has not yet contained the focus, the first element. Widgets that usually employ this pattern include grid and tree grid.

That's the reason why the focus could keep on the first message when the second message arrive.

However, if the user sent a message, we will consider the user has an intention to acknowledge the messages and we will take this opportunity to move the focus to the last message. E.g.

1. Assume the chat history has 1 message from the bot, e.g. greeting
2. The user move the focus to the chat history, it should focus on the greeting message
3. The user move the focus to the send box, then sent a message
4. The bot responded, thus, there are now 3 messages in the chat history (bot greeting, user message, bot respond)
5. The user move the focus to the chat history
6. The focus should be on the very last message in the chat history, instead of the first one (greeting)

It will be great if you can give a reference to the WAI-ARIA so we can understand our inconsistencies with the guideline.

**Comment by compulim** [comment]
@sknaappilaswd

When the user focus inside the message (e.g. "Button 1" in your first set of buttons), we intentionally trap the focus inside the message until the user press <kbd>ESC</kbd> key.

Web Chat is a complex widget (a.k.a. widget could contains more widgets), and we are implement the navigation pattern based on [layout grid navigation pattern](https://www.w3.org/TR/wai-aria-practices-1.2/#gridNav_inside). Excerpt from the WAI-ARIA Authoring Practices guideline:

> - `Tab`: moves focus to the next widget in the grid. Optionally, the focus movement may wrap inside a single cell or within the grid itself.
> - `Shift` + `Tab`: moves focus to the previous widget in the grid. Optionally, the focus movement may wrap inside a single cell or within the grid itself.

We implemented the optional part of the pattern and wrap the focus movement inside the cell. When the user press `Enter` key to "focus inside" the message, the focus will be wrapped around until the user press `ESC` key to restore the navigation pattern.

**Comment by sknaappilaswd** [comment]
@compulim 

Hi! There might be a misunderstanding. I see that you refer to multiple messages, but I am talking about one message (bot's answer to the user) that contains multiple sets of buttons. I am not able to get to the second set of buttons because the focus stays in the first set of buttons. If I press ESC the focus leaves the message and I may navigate to other messages, but I do not want to leave the bot's message but instead be able to navigate within it.

WCAG 1.3.2: Meaningful Sequence (A), WCAG 2.4.3: Focus Order (A).

Here is the AdaptiveCard that was used to reproduce this. The user asks a question and the bot answers with this:

`{
  "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
  "type": "AdaptiveCard",
  "version": "1.3",
  "body": [
    {
      "type": "TextBlock",
      "text": "An ActionSet in the body:",
      "wrap": true,
      "weight": "bolder"
    },
    {
      "type": "ActionSet",
      "actions": [
          {
            "type": "Action.Submit",
            "title": "Button 1",
            "wrap": true,
            "data": {
              "continue": "true"
            }
          },
          {
            "type": "Action.Submit",
            "title": "Button 2",
            "wrap": true,
            "data": {
              "continue": "true"
            }
          },
          {
            "type": "Action.Submit",
            "title": "Button 3",
            "wrap": true,
            "data": {
              "continue": "true"
            }
          }
      ]
    },
    {
      "type": "TextBlock",
      "text": "An ActionSet in the body (buttons not accessible by keyboard. If buttons are accessed using TAB, this text is not read to to screenreader):",
      "wrap": true,
      "weight": "bolder",
      "spacing": "extraLarge"
    },
    {
      "type": "ActionSet",
      "actions": [
          {
            "type": "Action.Submit",
            "title": "Button 1",
            "wrap": true,
            "data": {
              "continue": "true"
            }
          },
          {
            "type": "Action.Submit",
            "title": "Button 2",
            "wrap": true,
            "data": {
              "continue": "true"
            }
          }
      ]
    }
  ]
}`

**Comment by tdurnford** [comment]
@compulim Any updates on this issue?

**Comment by msomanathan** [comment]
@compulim Could you please provide updates on this issue? Thanks!

**Comment by axelsrz** [comment]
@compulim any updates?

**Comment by compulim** [comment]
I am using the Adaptive Card from @sknaappilaswd and created a chat history (attached below).

When I load it into Web Chat 4.15.2, I can press <kbd>TAB</kbd> and <kbd>SHIFT</kbd> + <kbd>TAB</kbd> key to focus on all 5 buttons of both sets. And NVDA 2022-01 did narrate all the buttons correctly when I am using <kbd>TAB</kbd> key to focus between them.

<details>
<summary>Click to see the chat history JSON</summary>

```json
[
  {
    "id": "a-00001",
    "text": "What can I do for you?",
    "timestamp": "2022-06-09T21:18:06.367Z",
    "type": "message",
    "attachments": [
      {
        "contentType": "application/vnd.microsoft.card.adaptive",
        "content": {
          "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
          "type": "AdaptiveCard",
          "version": "1.3",
          "body": [
            {
              "type": "TextBlock",
              "text": "An ActionSet in the body:",
              "wrap": true,
              "weight": "bolder"
            },
            {
              "type": "ActionSet",
              "actions": [
                {
                  "type": "Action.Submit",
                  "title": "Button 1",
                  "wrap": true,
                  "data": {
                    "continue": "true"
                  }
                },
                {
                  "type": "Action.Submit",
                  "title": "Button 2",
                  "wrap": true,
                  "data": {
                    "continue": "true"
                  }
                },
                {
                  "type": "Action.Submit",
                  "title": "Button 3",
                  "wrap": true,
                  "data": {
                    "continue": "true"
                  }
                }
              ]
            },
            {
              "type": "TextBlock",
              "text": "An ActionSet in the body (buttons not accessible by keyboard. If buttons are accessed using TAB, this text is not read by the screenreader (NVDA)):",
              "wrap": true,
              "weight": "bolder",
              "spacing": "extraLarge"
            },
            {
              "type": "ActionSet",
              "actions": [
                {
                  "type": "Action.Submit",
                  "title": "Button 1",
                  "wrap": true,
                  "data": {
                    "continue": "true"
                  }
                },
                {
                  "type": "Action.Submit",
                  "title": "Button 2",
                  "wrap": true,
                  "data": {
                    "continue": "true"
                  }
                }
              ]
            }
          ]
        }
      }
    ]
  }
]
```

</details>

https://user-images.githubusercontent.com/1622400/172965749-3cb2177a-294e-4e35-9b34-a275b5bf2c9a.mp4

**Comment by sknaappilaswd** [comment]
But in this reproduction and video the text above the second set of buttons is never read, so the buttons miss their context. I have not been able to hear the text paragraphs of the second set of buttons. With the TAB key you can jump from one button to another by also jumping from one set to another, but without hearing instructions (e.g. a question) the button sets become meaningless.

Say you want to ask the user would they rather answer questions based on their age or shoe sizes. The first set of buttons (age) could contain buttons with "30-35", "36-40", 41-45". The second set of buttons could also contain buttons with the same or different values, but they would be shoe sizes. How can the screen reader user know what the numbers are about, if the text above the button set is not accessible?

That is what I meant with these WCAG 1.3.2: Meaningful Sequence (A), WCAG 2.4.3: Focus Order (A).

Indeed with the tab key you can focus on buttons in different sets, but then you miss the text content present in the card completely. I have only been able to access the text and buttons in the first button set (focus gets trapped when using arrow keys). Never been able to achieve the same with the second button set.

**Comment by compulim** [comment]
When pressing <kbd>TAB</kbd> keys to focus on different elements, non-interactive elements, such as text or `<p>`, are never read by the screen reader. Only interactive elements, such as `<button>` or `<a>`, will be narrated. This is behavior also true for normal HTML webpage.

Could you repro your Adaptive Card outside of Web Chat? You can paste your card JSON into [Adaptive Card Designer](https://adaptivecards.io/designer/), then turn on the "Preview mode" to test. This will help to determine whether this is an issue on Adaptive Cards or Web Chat.
